crates/wallet/src/wallet/mod.rs

```diff
@@ -372,6 +375,14 @@ impl Wallet {
         let genesis_hash = params
             .genesis_hash
             .unwrap_or(genesis_block(network).block_hash());
+
+        if genesis_hash.ne(&genesis_block(network).block_hash()) {
+            return Err(DescriptorError::Mismatch(MismatchError::Genesis {
+                network: genesis_block(network).block_hash(),
+                parameter: genesis_hash,
+            }));
+        }
+
```

Comment:
As mentioned by @ValuedMammal, the intention here is so the caller can provide a custom genesis hash.

However, if the intention is to enforce a 1-possible-genesis-hash-per-network-variant invariance, changing the data structure to disallow the invariance would make a lot more sense. I.e. remove `params.genesis_hash`.